### PR TITLE
Add otel

### PR DESCRIPTION
**Pull request by IgorEulalio**

### CODE REVIEW

File: actionners/kubernetes/drain/drain.go [hunk 1]
```diff
@@ -92,6 +92,7 @@ func Action(action *rules.Action, event *events.Event) (utils.LogLine, *model.Da
 
 	for _, p := range pods.Items {
 		wg.Add(1)
+		p := p // loopclosure: loop variable p captured by func literal
```
Comment (Issif) [inline actionners/kubernetes/drain/drain.go:95]: what's the point of that?

File: configuration/configuration.go [hunk 2]
```diff
@@ -10,19 +10,29 @@ import (
...
+	defaultOtelCollectorEndpoint        string = "localhost"
+	defaultOtelCollectorUseInsecureGrpc bool   = false
+	defaultOtelCollectorPort            int    = 4317
 )
 
+type Otel struct {
+	CollectorEndpoint        string `mapstructure:"collector_endpoint"`
+	CollectorPort            string `mapstructure:"collector_port"`
+	CollectorUseInsecureGrpc bool   `mapstructure:"collector_use_insecure_grpc"`
+}
```
Comment (Issif) [inline configuration/configuration.go:29]: The Endpoint should allow `ip:port` or `http(s):domain(:port)/path`, to work with the local collectors and SaaS (eg, Tempo of Grafana Cloud). It also means we should drop `collector_port` setting, cause it becomes useless.
For SaaS and other, we should allow to specify user/password, for now, I can't test with my Grafana Cloud account.

**Review by Issif — CHANGES_REQUESTED** [review]

**Review by Issif — COMMENTED** [review]
I saw some errors are not catched and printed outside the log function:


```
2024-06-30T15:22:41+02:00 INF match event="Test download" output=test priority=Critical rule="Test download" source=syscall trace_id=396b4e84291bcaa9f74a18f44fa21b6c

2024/06/30 15:22:55 traces export: context deadline exceeded: rpc error: code = Unavailable desc = name resolver error: produced zero addresses

2024-06-30T15:23:11+02:00 ERR action error="the pod 'falco-jj7sp' in the namespace 'falco' doesn't exist" action="Test log" actionner=kubernetes:log event=test rule="Test download" trace_id=396b4e84291bcaa9f74a18f44fa21b6c
```

File: actionners/kubernetes/drain/drain.go (hunk 1 above)
Comment (IgorEulalio) [reply actionners/kubernetes/drain/drain.go:95]: pleasing the linter

File: configuration/configuration.go (hunk 2 above)
Comment (IgorEulalio) [reply configuration/configuration.go:29]: I didn't know that it was possible to use a external collector instance, does that only work with Grafana cloud? I'll try create a free account and change this.

**Comment by IgorEulalio** [comment]
> I saw some errors are not catched and printed outside the log function:
> 
> ```
> 2024-06-30T15:22:41+02:00 INF match event="Test download" output=test priority=Critical rule="Test download" source=syscall trace_id=396b4e84291bcaa9f74a18f44fa21b6c
> 
> 2024/06/30 15:22:55 traces export: context deadline exceeded: rpc error: code = Unavailable desc = name resolver error: produced zero addresses
> 
> 2024-06-30T15:23:11+02:00 ERR action error="the pod 'falco-jj7sp' in the namespace 'falco' doesn't exist" action="Test log" actionner=kubernetes:log event=test rule="Test download" trace_id=396b4e84291bcaa9f74a18f44fa21b6c
> ```

That is expected behavior if OTEL is not configured, I've configured app to lazy init OTEL forwarder and whenever we generate the first trace we try to communicate with collector and send it. 

We can have a flag otel.enabled which defaults to false to prevent this if you think it's better.

File: cmd/server.go [hunk 3]
```diff
@@ -247,19 +249,44 @@ var serverCmd = &cobra.Command{
...
+		}()
+
 		if err := srv.ListenAndServe(); err != nil {
 			utils.PrintLog("fatal", utils.LogLine{Error: err.Error(), Message: "http"})
 		}
 	},
 }
 
+func newHTTPHandler() http.Handler {
+	mux := http.NewServeMux()
```
Comment (Issif) [inline cmd/server.go:275]:
```suggestion
	mux := http.NewServeMux()
	mux.Handle("/metrics", metrics.Handler())
```

File: cmd/server.go [hunk 4]
```diff
@@ -247,19 +249,44 @@ var serverCmd = &cobra.Command{
...
+
+	handleFunc := func(pattern string, handlerFunc func(http.ResponseWriter, *http.Request)) {
+		otelHandler := otelhttp.WithRouteTag(pattern, http.HandlerFunc(handlerFunc))
+		mux.Handle(pattern, otelHandler)
+	}
+
+	handleFunc("/", handler.MainHandler)
+	handleFunc("/healthz", handler.HealthHandler)
+	handleFunc("/rules", handler.RulesHandler)
```
Comment (Issif) [inline cmd/server.go:284]: Does it mean, the traces are also created for these endpoints?

**Review by Issif — CHANGES_REQUESTED** [review]

File: cmd/server.go [hunk 5]
```diff
@@ -247,19 +249,44 @@ var serverCmd = &cobra.Command{
...
 		if err != nil {
 			utils.PrintLog("fatal", utils.LogLine{Error: err.Error(), Message: "nats"})
 		}
 		go actionners.StartConsumer(c)
 
 		utils.PrintLog("info", utils.LogLine{Result: fmt.Sprintf("Falco Talon is up and listening on %s:%d", config.ListenAddress, config.ListenPort), Message: "http"})
 
+		otelShutdown, err := tracing.SetupOTelSDK(context.Background())
+		if err != nil {
+			utils.PrintLog("warn", utils.LogLine{Error: err.Error(), Message: "fail to initialize OTEL GRPC exporter."})
```
Comment (Issif) [inline cmd/server.go:262]:
```suggestion
			utils.PrintLog("warn", utils.LogLine{Error: err.Error(), Message: "fail to initialize OTEL gRPC exporter."})
```

File: cmd/server.go [hunk 6]
```diff
@@ -247,19 +249,44 @@ var serverCmd = &cobra.Command{
...
+		}
+		defer func() {
+			err = errors.Join(err, otelShutdown(context.Background()))
```
Comment (Issif) [inline cmd/server.go:265]: I don't see where this error is used. The parent func doesn't return any error.

File: cmd/server.go (hunk 5 above)
Comment (Issif) [inline cmd/server.go:262]:
```suggestion
			utils.PrintLog("warn", utils.LogLine{Error: err.Error(), Message: "otel-traces"})
```

Comment (Issif) [reply cmd/server.go:262]: The error should be contained in the Error

**Review by Issif — CHANGES_REQUESTED** [review]

File: cmd/server.go (hunk 6 above)
Comment (Issif) [reply cmd/server.go]: Still don't understand

File: actionners/actionners.go [hunk 7]
```diff
@@ -25,12 +31,14 @@ import (
 	aws "github.com/falco-talon/falco-talon/internal/aws/client"
 	calico "github.com/falco-talon/falco-talon/internal/calico/client"
 	cilium "github.com/falco-talon/falco-talon/internal/cilium/client"
-	"github.com/falco-talon/falco-talon/internal/context"
+	falcoContext "github.com/falco-talon/falco-talon/internal/context"
```
Comment (Issif) [inline actionners/actionners.go:34]:
```suggestion
	talonContext "github.com/falco-talon/falco-talon/internal/context"
```

wdyt?

**Pull request merged**